src/main.cpp

```diff
@@ -11,124 +11,126 @@
+		  256 * std::clamp(sycl::sqrt(fb_data[pixel_index].y()), 0.0f, 0.999f));
+	  int b = static_cast<int>(
+		  256 * std::clamp(sycl::sqrt(fb_data[pixel_index].z()), 0.0f, 0.999f));
+	  std::cout << r << " " << g << " " << b << "\n";
+	}
   }
 }
 
 int main() {
   // Frame buffer dimensions
-  constexpr auto width = 800;
-  constexpr auto height = 480;
+  constexpr auto width = buildparams::output_width;
+  constexpr auto height = buildparams::output_height;
 
   // Allocate frame buffer on host
-  std::vector<color> fb(width * height);
+  std::array<color, (width * height)> fb;
 
   /// Graphical objects
   std::vector<hittable_t> hittables;
 
   // Generating a checkered ground and some random spheres
   texture_t t =
-      checker_texture(color { 0.2f, 0.3f, 0.1f }, color { 0.9f, 0.9f, 0.9f });
+	  checker_texture(color { 0.2f, 0.3f, 0.1f }, color { 0.9f, 0.9f, 0.9f });
   material_t m = lambertian_material(t);
   hittables.emplace_back(sphere(point { 0, -1000, 0 }, 1000, m));
   t = checker_texture(color { 0.9f, 0.9f, 0.9f }, color { 0.4f, 0.2f, 0.1f });
 
+  LocalPseudoRNG rng;
+
   for (int a = -11; a < 11; a++) {
-    for (int b = -11; b < 11; b++) {
-      // Based on a random variable , the material type is chosen
-      auto choose_mat = random_float();
-      // Spheres are placed at a point randomly displaced from a,b
-      point center(a + 0.9f * random_float(), 0.2f, b + 0.9f * random_float());
-      if (sycl::length((center - point(4, 0.2f, 0))) > 0.9f) {
-        if (choose_mat < 0.4f) {
-          // Lambertian
-          auto albedo = randomvec() * randomvec();
-          hittables.emplace_back(
-              sphere(center, 0.2f, lambertian_material(albedo)));
-        } else if (choose_mat < 0.8f) {
-          // Lambertian movig spheres
-          auto albedo = randomvec() * randomvec();
-          auto center2 = center + point { 0, random_float(0, 0.25f), 0 };
-          hittables.emplace_back(sphere(center, center2, 0.0f, 1.0f, 0.2f,
-                                        lambertian_material(albedo)));
-        } else if (choose_mat < 0.95f) {
-          // metal
-          auto albedo = randomvec(0.5f, 1);
-          auto fuzz = random_float(0, 0.5f);
-          hittables.emplace_back(
-              sphere(center, 0.2f, metal_material(albedo, fuzz)));
-        } else {
-          // glass
-          hittables.emplace_back(
-              sphere(center, 0.2f,
-                     dielectric_material(1.5f, color { 1.0f, 1.0f, 1.0f })));
-        }
-      }
-    }
+	for (int b = -11; b < 11; b++) {
+	  // Based on a random variable , the material type is chosen
+	  auto choose_mat = random_float(rng);
+	  // Spheres are placed at a point randomly displaced from a,b
+	  point center(a + 0.9f * random_float(rng), 0.2f, b + 0.9f * random_float(rng));
+	  if (sycl::length((center - point(4, 0.2f, 0))) > 0.9f) {
+		if (choose_mat < 0.4f) {
+		  // Lambertian
+		  auto albedo = randomvec(rng) * randomvec(rng);
```

Comment:
Curious this factor of 2 random variables, by the way...
This is why I hate uncommented code. :-)
But of course you are not the culprit.
Perhaps you can find why by looking at the blog?

Reply:
https://raytracing.github.io/books/RayTracingInOneWeekend.html#wherenext?/afinalrender 
It's also use a product of random vector without further explanation.
 
I guess it add a bias toward vector with lower components  (randomvec gives a vector with coordinates in [0, 1], so the product of two randomvec() is also in [0,1] but when with one randomvec you get a component below 0.5 with probability 0.5, with a product of randomvec the probability is 0.75)

Reply:
Should it be `auto albedo = randomvec(rng) * randomvec(rng);` or
```c++
auto r = randomvec(rng);
auto albedo = r * r;
```
would be enough?

Reply:
The two options are not equivalent (on this the probability of a coordinate below 0.5 is around 0.707 for instance) , but I guess it should be sufficient. 
However, this line is from the generation of the object in the main and not in the rendering kernel, so I am not sure there is a performance issue here.